flang/include/flang/Lower/PFTBuilder.h

```diff
@@ -327,13 +327,21 @@ struct Variable {
 
   const Fortran::semantics::Symbol &getSymbol() const { return *sym; }
   bool isGlobal() const { return global; }
+  bool isHeapAlloc() const { return heapAlloc; }
+  bool isPointer() const { return pointer; }
+  bool isTarget() const { return target; }
   int getDepth() const { return depth; }
+  void setHeapAlloc(bool to = true) { heapAlloc = to; }
+  void setPointer(bool to = true) { pointer = to; }
+  void setTarget(bool to = true) { target = to; }
 
 private:
   const Fortran::semantics::Symbol *sym;
   int depth;
   bool global;
-  //bool heap{false}; // variable needs deallocation on exit
+  bool heapAlloc{false}; // variable needs deallocation on exit
+  bool pointer{false};
+  bool target{false};
```

Comment:
I'll add a better comment.

The motivation is that lowering may want to promote variables beyond the syntactic look. For example, if someone writes
```
   INTEGER :: my_array(1000,1000,10)
```
lowering may decide to lower that variable as a heap allocation to avoid blowing the stack. That is ok, but lowering cannot expect semantics to materialize an "ALLOCATABLE" attribute in that case.

Reply:
There are other similar issues, I suspect, such as the actual layout of arrays in parameterized derived types (where arrays that depend on the size parameter will be implemented as allocatables).    Do we need to carry such information in module files?

Reply:
The type of transformation (https://github.com/flang-compiler/f18-llvm-project/pull/101#discussion_r431362854) would be confined to the scope of a subprogram.

In the case of observable layout decisions, lowering should see the same Fortran derived type (from a consistent module file) and lower it using the same process. That could be up to and including calling some Fortran runtime helpers to determine the layout, etc.